common/djangoapps/third_party_auth/saml.py

```diff
@@ -93,3 +95,158 @@ def wrapped_method(*args, **kwargs):
     def _config(self):
         from .models import SAMLConfiguration
         return SAMLConfiguration.current(Site.objects.get_current(get_current_request()))
+
+
+class SapSuccessFactorsIdentityProvider(SAMLIdentityProvider):
```

Comment:
@haikuginger If you sketch it out, it does require significantly less code with the approach I'm advocating, and there's nothing inherently worse about passing an object via a method parameter vs. accessing it on an object using `self` - in fact the method method (haha) makes it more explicit what information is being used where.

Not only does this approach use less code and re-use existing settings fields/patterns, but if you look at the current backend model vs. the SAMLIdentityProvider class, you can see that the SAMLIdentityProvider class is currently just a dumb data store and none of its methods have side effects, unlike the backend class which is meant to be the class that "does things" and whose methods already have side effects and do things like spawning of HTTP requests.

Sketch:
```python
class CouldNotRetrieveDetailsException(Exception):
    pass


class SapSuccessFactorsBackend(SAMLAuthBackend):

    def get_user_details(self, response):
        """ Get user details like full name, email, etc. from the response - see auth_complete """
        idp = self.get_idp(response['idp_name'])
        details = idp.get_user_details(response['attributes'])
        try:
            details = self.get_sap_user_details(details.get('username', idp.conf)
        except CouldNotRetrieveDetailsException:
            pass
        return details
    
    def get_sap_user_details(self, username, idp_conf):
        try:
            sapsf_idp_url = idp_conf['sapsf_oauth_root_url'] + 'idp'
            sapsf_token_url = idp_conf['sapsf_oauth_root_url'] + 'token'
            sapsf_private_key = return idp_conf['sapsf_private_key']
            ...
        except KeyError as exc:
            log.warning(
                "To retrieve rich user data for an SAP SuccessFactors identity provider, the following key in "
                "'other_settings' is required, but was missing: %s",
                exc.args[0]
            )
            raise CouldNotRetrieveDetailsException
        
        try:
            client = self.get_odata_api_client(username, odata_company_id, odata_client_id, sapsf_token_url, sapsf_private_key)
            response = client.get(
                '{root_url}User(userId=\'{user_id}\')?$select=username,firstName,lastName,defaultFullName,email'.format(
                    root_url=self.odata_api_root_url,
                    user_id=username
                ),
                timeout=10,
            )
            response.raise_for_status()
            response = response.json()
        except requests.RequestException:
            log.warning(
                'Unable to retrieve user details with username %s from SAPSuccessFactors with company ID %s.',
                username,
                self.odata_company_id,
            )
            raise CouldNotRetrieveDetailsException

        return {
            'username': response['d']['username'],
            'first_name': response['d']['firstName'],
            'last_name': response['d']['lastName'],
            'fullname': response['d']['defaultFullName'],
            'email': response['d']['email'],
        }

    def get_odata_api_client(username, odata_company_id, odata_client_id, sapsf_token_url, sapsf_private_key):
        ...
```


What do you think?